### ISSUE TITLE
fetchpatch may produce broken patches

### ISSUE THREAD

**Issue by volth** [NixOS/nixpkgs#37375]
## Issue description

There is a function ```fetchpatch``` which is like ```fetchurl``` following some "normalization".
That "normalization" is not correct.

Example

1. Input, two-step patch, on the first step a new file is created from scratch, on the second step the file is patched: https://github.com/NixOS/nixpkgs/compare/master...volth:processing3.patch

2. code 
```
fetchpatch {
  name   = "processing3";
  url    = "https://github.com/NixOS/nixpkgs/compare/master...volth:processing3.patch";
  sha256 = "07iiawwyi33rnblhvij78igryvgvxki94ck4amxp75r4dzsd5kzd";
}
```

3. output: https://gist.github.com/volth/9eb782d69c33c6186509398b29b0a65c
The chunks are reordered so on the first step non-existent file is patched.
Applying this patch would fail.

**Comment by volth** [comment]
That seems easy to fix but the fix likely changes all the ```sha256``` of ```fetchpatch```

**Comment by nicknovitski** [comment]
Maybe add an optional argument which enables the fix?

**Comment by vcunat** [comment]
Hmm, probably.  The main point of `fetchpatch` is to avoid hash changes for "unstable" patches (dynamically generated).

**Comment by vcunat** [comment]
It would be better, but I'm not keen on breaking hundreds of hashes in nixpkgs (which we perhaps can replace by some clever script) and unknown number of hashes in user expression (which we can't fix).

**Comment by vcunat** [comment]
Yes, that would be nice, though possibly a little messy to write the implementation.

**Comment by danbst** [comment]
Another example when `fetchpatch` did wrong reorder (edit first, create next):
```
nix-build -E 'with import <nixpkgs>{}; fetchpatch {
   url = "https://github.com/NixOS/nixpkgs/pull/18419.patch";
   sha256 = "0ggcqyffhfl3p1lnjgcl6knqairdfn7r5mrizrlanpmlgdgk90s7";
}'
```

**Comment by danbst** [comment]
As for me, I'd be fine with extra argument, which fixes the problem. The alternative for me was to use fetchurl and `.diff` Github patch, but it did change hash every 2 months or so.
